### PR TITLE
refactor: media preview

### PR DESCRIPTION
**Pull request by dudu0506**
# Pull Request

## Related Issue

Closes FW-0000

## Checklist

-   [ ] I have tested these changes locally.
-   [ ] I have reviewed the code changes.
-   [ ] I have added unit tests if applicable.

### CODE REVIEW

File: src/components/Markup/ArticleMarkup.tsx [hunk 1]
```diff
@@ -65,9 +66,10 @@ export const ArticleMarkup = memo<ArticleMarkupProps>(function ArticleMarkup({
                                 event.preventDefault();
                                 event.stopPropagation();
                                 if (!props.src) return;
-                                PreviewImageModalRef.open({
-                                    current: props.src,
-                                    images: images.current,
+                                PreviewMediaModalRef.open({
+                                    index: `${index > -1 ? index : 0}`,
```
Comment (nuanyang233) [inline src/components/Markup/ArticleMarkup.tsx:70]: Math.min

File: src/app/(normal)/article/[id]/pages/DetailPage.tsx [hunk 2]
```diff
@@ -91,9 +92,10 @@ export function ArticleDetailPage({ params: { id: articleId } }: PageProps) {
                             event.stopPropagation();
                             event.preventDefault();
                             if (cover.data)
-                                PreviewImageModalRef.open({
-                                    images: [cover.data],
-                                    current: cover.data,
+                                PreviewMediaModalRef.open({
+                                    medias: [{ type: 'Image', uri: cover.data }],
+                                    index: '1',
```
Comment (guanbinrui) [inline src/app/(normal)/article/[id]/pages/DetailPage.tsx:97]: Start from 0.

File: src/components/PreviewMedia/index.tsx [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+import { CloseButton } from '@/components/CloseButton.js';
+import { Modal } from '@/components/Modal.js';
+import { PreviewContent } from '@/components/PreviewMedia/PreviewContent.js';
+import type { Source } from '@/constants/enum.js';
+import { EMPTY_LIST, SUPPORTED_PREVIEW_MEDIA_TYPES } from '@/constants/index.js';
+import { useIsMedium } from '@/hooks/useMediaQuery.js';
+import type { Attachment, Post } from '@/providers/types/SocialMedia.js';
+
+interface PreviewMediaProps {
+    index: string;
```
Comment (guanbinrui) [inline src/components/PreviewMedia/index.tsx:22]:
```suggestion
    index: number;
```

File: src/components/PreviewMedia/index.tsx [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+                            }
+                        }}
+                        keyboard
+                        initialSlide={Number(index) - 1}
+                    >
+                        {images.map((asset, key) => {
+                            return (
+                                <SwiperSlide key={key} className="flex">
+                                    <div className="flex h-full w-full items-center justify-center">
+                                        <PreviewContent source={source!} asset={asset} />
```
Comment (guanbinrui) [inline src/components/PreviewMedia/index.tsx:79]: source could be undefined.

File: src/components/PreviewMedia/index.tsx [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+                                    </div>
+                                </SwiperSlide>
+                            );
+                        })}
+                        <ClickableButton
+                            ref={prevRef}
+                            className="prev-button absolute left-[50px] top-[50%] z-[9999] max-md:hidden"
```
Comment (guanbinrui) [inline src/components/PreviewMedia/index.tsx:86]: Please use tailwind recommended z-indexes.

File: src/components/PreviewMedia/index.tsx [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+                        >
+                            <ArrowLeftIcon width={24} height={24} className="rounded-full p-1 text-main hover:bg-bg" />
+                        </ClickableButton>
+                        <ClickableButton
+                            ref={nextRef}
+                            className="next-button absolute right-[50px] top-[50%] z-[9999] max-md:hidden"
```
Comment (guanbinrui) [inline src/components/PreviewMedia/index.tsx:92]: Please use tailwind recommended z-indexes.

File: src/components/PreviewMedia/index.tsx [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+        const asset = post.metadata.content?.asset;
+        const imageAttachments =
+            post.metadata.content?.attachments?.filter((x) => SUPPORTED_PREVIEW_MEDIA_TYPES.includes(x.type)) ??
+            EMPTY_LIST;
+
+        if (asset?.type === 'Image' && imageAttachments.length === 1) {
+            return [{ ...asset }];
+        }
+        return imageAttachments;
+    }, [post, medias]);
```
Comment (guanbinrui) [inline src/components/PreviewMedia/index.tsx:48]:
```suggestion
    const images = useMemo(() => {
        if (!post) return EMPTY_LIST;
        const asset = post.metadata.content?.asset;
        const imageAttachments =
            post.metadata.content?.attachments?.filter((x) => SUPPORTED_PREVIEW_MEDIA_TYPES.includes(x.type)) ??
            EMPTY_LIST;

        if (asset?.type === 'Image' && imageAttachments.length === 1) {
            return [{ ...asset }];
        }
        return imageAttachments;
    }, [post]);
```

File: src/components/PreviewMedia/index.tsx [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+                        onBeforeInit={(swiper) => {
+                            if (typeof swiper.params.navigation === 'object') {
+                                swiper.params.navigation.prevEl = prevRef.current;
+                                swiper.params.navigation.nextEl = nextRef.current;
+                            }
+                        }}
+                        keyboard
+                        initialSlide={Number(index) - 1}
+                    >
+                        {images.map((asset, key) => {
```
Comment (guanbinrui) [inline src/components/PreviewMedia/index.tsx:75]:
```suggestion
                        {(medias || images).map((asset, key) => {
```

File: src/components/Markup/ArticleMarkup.tsx [hunk 9]
```diff
@@ -65,9 +66,10 @@ export const ArticleMarkup = memo<ArticleMarkupProps>(function ArticleMarkup({
                                 event.preventDefault();
                                 event.stopPropagation();
                                 if (!props.src) return;
-                                PreviewImageModalRef.open({
-                                    current: props.src,
-                                    images: images.current,
+                                PreviewMediaModalRef.open({
+                                    index: `${Math.min(index + 1, 1)}`,
```
Comment (swkatmask) [inline src/components/Markup/ArticleMarkup.tsx:70]: Do you indent?
```suggestion
                                    index: `${Math.max(index + 1, 1)}`,
```

File: src/components/PreviewMedia/index.tsx [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    showAction?: boolean;
+    onClose: () => void;
+}
+
+export function PreviewMedia({ post, source, medias, index, open, showAction = true, onClose }: PreviewMediaProps) {
+    const prevRef = useRef<HTMLButtonElement>(null);
+    const nextRef = useRef<HTMLButtonElement>(null);
+    const isMedium = useIsMedium();
+
+    const images = useMemo(() => {
```
Comment (guanbinrui) [inline src/components/PreviewMedia/index.tsx:36]:
```suggestion
    const assets = useMemo(() => {
```

File: src/components/PreviewMedia/index.tsx [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+import { PreviewContent } from '@/components/PreviewMedia/PreviewContent.js';
+import type { Source } from '@/constants/enum.js';
+import { EMPTY_LIST, SUPPORTED_PREVIEW_MEDIA_TYPES } from '@/constants/index.js';
+import { useIsMedium } from '@/hooks/useMediaQuery.js';
+import type { Attachment, Post } from '@/providers/types/SocialMedia.js';
+
+interface PreviewMediaProps {
+    index: string;
+    open: boolean;
+    source?: Source;
```
Comment (guanbinrui) [inline src/components/PreviewMedia/index.tsx:24]:
```suggestion
    source: Source;
```

File: src/modals/PreviewMediaModal.tsx [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+                setSource(props.source);
+                setShowAction(props.showAction);
+            },
+            onClose: () => {
+                setMedias(undefined);
+                setPost(undefined);
+                setShowAction(false);
+                setIndex('');
+            },
+        });
```
Comment (guanbinrui) [inline src/modals/PreviewMediaModal.tsx:38]: const [props, setProps] = useState<PreviewMediaProps>(null)

onOpen: (props) => {
    setProps(props)
},

onClose: () => {
    setProps(null)
}

if (!props) return null

Comment (guanbinrui) [reply src/modals/PreviewMediaModal.tsx:38]: Might working for `src/modals/ConfirmModal.tsx` as well.

File: src/modals/PreviewMediaModal.tsx [hunk 13]
```diff
@@ -0,0 +1,51 @@
+import { forwardRef, useState } from 'react';
+
+import { PreviewMedia } from '@/components/PreviewMedia/index.js';
+import type { Source } from '@/constants/enum.js';
+import { useSingletonModal } from '@/hooks/useSingletonModal.js';
+import type { SingletonModalRefCreator } from '@/libs/SingletonModal.js';
+import type { Attachment, Post } from '@/providers/types/SocialMedia.js';
+
+export interface PreviewMediaModalOpenProps {
```
Comment (guanbinrui) [inline src/modals/PreviewMediaModal.tsx:9]:
```suggestion
export interface PreviewMediaModalOpenProps extends PreviewMediaProps {
```

File: src/app/(normal)/@modal/(.)post/[id]/photos/[index]/page.tsx [hunk 14]
```diff
@@ -74,70 +39,5 @@ export default function PreviewPhotoModal({ params: { id: postId, index }, searc
...
-                        </ClickableButton>
-                    </Swiper>
-                </div>
-                <div className="absolute my-1 flex items-center justify-between bottom-safe">
-                    {post ? <PostActionsWithGrid className="gap-8" post={post} disablePadding /> : null}
-                </div>
-            </div>
-        </Modal>
-    );
+    return <PreviewMedia open post={post} source={currentSource} index={index} onClose={() => router.back()} />;
```
Comment (swkatmask) [inline src/app/(normal)/@modal/(.)post/[id]/photos/[index]/page.tsx:42]: Suggestion: separate `offset` and `index` from UI input, you can use `offset` as 0-indexed confidently inside `<PreviewMedia />`, besides, the UI output, don't forget to plus 1 to convert `offset` to `index`

```suggestion
    return <PreviewMedia open post={post} source={currentSource} offset={parse(index, 10)} onClose={() => router.back()} />;
```

File: src/components/PreviewMedia/index.tsx [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+    onClose: () => void;
+}
+
+export function PreviewMedia({ post, source, medias, index, open, showAction = true, onClose }: PreviewMediaProps) {
+    const prevRef = useRef<HTMLButtonElement>(null);
+    const nextRef = useRef<HTMLButtonElement>(null);
+    const isMedium = useIsMedium();
+
+    const images = useMemo(() => {
+        if (medias) return [...medias];
```
Comment (swkatmask) [inline src/components/PreviewMedia/index.tsx:37]: You won't mutate it, right? No need to create a copy.
```suggestion
        if (medias) return medias;
```

File: src/components/PreviewMedia/index.tsx [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+        if (!post) return EMPTY_LIST;
+        const asset = post.metadata.content?.asset;
+        const imageAttachments =
+            post.metadata.content?.attachments?.filter((x) => SUPPORTED_PREVIEW_MEDIA_TYPES.includes(x.type)) ??
+            EMPTY_LIST;
+
+        if (asset?.type === 'Image' && imageAttachments.length === 1) {
+            return [{ ...asset }];
```
Comment (swkatmask) [inline src/components/PreviewMedia/index.tsx:45]: ditto

File: src/components/Posts/VideoAsset.tsx [hunk 17]
```diff
@@ -36,9 +37,13 @@ export function VideoAsset({ asset, isQuote, source }: VideoAssetProps) {
             ) : null}
         </div>
     ) : (
-        <Video src={source === Source.Twitter ? forwardTwitterVideo(asset.uri) : asset.uri} poster={asset.coverUri}>
+        <Video
+            autoPlay={autoPlay}
+            src={source === Source.Twitter ? forwardTwitterVideo(asset.uri) : asset.uri}
+            poster={asset.coverUri}
+        >
             {asset.type === 'AnimatedGif' ? (
```
[the remark this answers is not in the document]
Comment (dudu0506) [reply src/components/Posts/VideoAsset.tsx:45]: Yes. The video player we use will append some query in url, and this will be blocked by twitter

**Review by guanbinrui — APPROVED** [review]

**Pull request merged**